### ISSUE TITLE
Conditional Generation Evaluation Request

### ISSUE THREAD

**Issue by CoderChen01** [ML-GSAI/LLaDA#42]
Dear Authors,

Thank you for sharing your insights on this exciting project.

We noticed that you have already provided the `loglikelihood` implementation for lm_eval, but the existing code has not yet included `generate_until`, making it impossible to evaluate the model's conditional generation capabilities.

Do you have plans to make this feature available?

Best regards,  
Jorji

**Comment by Monohydroxides** [comment]
Since the lm-evaluation-harness does not support the evaluation of certain metrics (e.g., HumanEval-FIM), we have utilized an internal evaluation library to assess conditional generation metrics. We currently have no plans to open-source this portion of the code.

**Comment by CoderChen01** [comment]
Thank you for your reply. Looking forward to your further work.

**Comment by CoderChen01** [comment]
Dear @Monohydroxides,

I would like to know if your internal evaluation library supports batch processing for conditional generation evaluation. Additionally, could you specify which GPU is used for this process? What is the maximum duration for an evaluation on a benchmark? Thank you for taking the time to answer my questions.

Best regards,
Jorji

**Comment by f14-bertolotti** [comment]
**Hi @Monohydroxides,**  

First off, great work!  

I'm trying to reproduce the results from your paper using `lm-evaluation-harness`, specifically focusing on **Table 6 (page 19)** with **"Randomly remasking"** and **LLaDA 8B Base** (which has a reported value of **52.3**).  

However, when I run my implementation, I only obtain a score of around **20** using `lm-evaluation-harness`.  

My code is based on your [`generate.py`](https://github.com/ML-GSAI/LLaDA/blob/main/generate.py). Below is a minimal version that should be reproducible as long as `generate.py` is available:  


```python
from generate import generate
import transformers, tqdm, pprint, lm_eval, torch

class RandomMaskingLLaDA(lm_eval.api.model.LM):
    def __init__(self):
        super().__init__()

        self.model = transformers.AutoModelForCausalLM.from_pretrained(
            "GSAI-ML/LLaDA-8B-Base", 
            trust_remote_code=True, 
            torch_dtype=torch.bfloat16, 
        ).to("cuda:0")
        self.tokenizer = transformers.AutoTokenizer.from_pretrained(
            "GSAI-ML/LLaDA-8B-Base", 
            trust_remote_code=True
        )

        self.device = "cuda:0"
        self.mask_id = 126336
        self.model.eval()

    def loglikelihood(*args, **kwargs): raise NotImplemented() 
    def loglikelihood_rolling(*args, **kwargs): raise NotImplemented() 

    @torch.no_grad()
    def generate_until(self, requests):
        results = []
        for i,request in enumerate(bar:=tqdm.tqdm(requests)):
            bar.set_description(f"{i}")
            source_ten = torch.tensor(self.tokenizer(request.args[0])["input_ids"]).unsqueeze(0)

            result = generate(
                self.model,
                prompt=source_ten,
                steps=256,
                gen_length=512,
                block_length=512,
                temperature=0.0,
                cfg_scale=0.0,
                remasking="random",
                mask_id=126336,
            )
            results.append(self.tokenizer.decode(result[0,-512:]))

        return results

if __name__ == "__main__":
    results = lm_eval.simple_evaluate(
        model = RandomMaskingLLaDA(),
        tasks=["gsm8k"],
        limit = None, # no. requests
        num_fewshot = 4,
    )
    with open("results.txt", "w") as f:
        pprint.pprint(results, stream=f)

```
I completely understand and respect your decision not to share code related to generation. However, could you provide more details on how generation benchmarks (e.g., `gsm8k`) are evaluated? Any insights would be greatly appreciated!  

Thanks in advance!

**Comment by CoderChen01** [comment]
> **Hi [@Monohydroxides](https://github.com/Monohydroxides),**
> 
> First off, great work!
> 
> I'm trying to reproduce the results from your paper using `lm-evaluation-harness`, specifically focusing on **Table 6 (page 19)** with **"Randomly remasking"** and **LLaDA 8B Base** (which has a reported value of **52.3**).
> 
> However, when I run my implementation, I only obtain a score of around **20** using `lm-evaluation-harness`.
> 
> My code is based on your [`generate.py`](https://github.com/ML-GSAI/LLaDA/blob/main/generate.py). Below is a minimal version that should be reproducible as long as `generate.py` is available:
> 
> from generate import generate
> import transformers, tqdm, pprint, lm_eval, torch
> 
> class RandomMaskingLLaDA(lm_eval.api.model.LM):
>     def __init__(self):
>         super().__init__()
> 
>         self.model = transformers.AutoModelForCausalLM.from_pretrained(
>             "GSAI-ML/LLaDA-8B-Base", 
>             trust_remote_code=True, 
>             torch_dtype=torch.bfloat16, 
>         ).to("cuda:0")
>         self.tokenizer = transformers.AutoTokenizer.from_pretrained(
>             "GSAI-ML/LLaDA-8B-Base", 
>             trust_remote_code=True
>         )
> 
>         self.device = "cuda:0"
>         self.mask_id = 126336
>         self.model.eval()
> 
>     def loglikelihood(*args, **kwargs): raise NotImplemented() 
>     def loglikelihood_rolling(*args, **kwargs): raise NotImplemented() 
> 
>     @torch.no_grad()
>     def generate_until(self, requests):
>         results = []
>         for i,request in enumerate(bar:=tqdm.tqdm(requests)):
>             bar.set_description(f"{i}")
>             source_ten = torch.tensor(self.tokenizer(request.args[0])["input_ids"]).unsqueeze(0)
> 
>             result = generate(
>                 self.model,
>                 prompt=source_ten,
>                 steps=256,
>                 gen_length=512,
>                 block_length=512,
>                 temperature=0.0,
>                 cfg_scale=0.0,
>                 remasking="random",
>                 mask_id=126336,
>             )
>             results.append(self.tokenizer.decode(result[0,-512:]))
> 
>         return results
> 
> if __name__ == "__main__":
>     results = lm_eval.simple_evaluate(
>         model = RandomMaskingLLaDA(),
>         tasks=["gsm8k"],
>         limit = None, # no. requests
>         num_fewshot = 4,
>     )
>     with open("results.txt", "w") as f:
>         pprint.pprint(results, stream=f)
> I completely understand and respect your decision not to share code related to generation. However, could you provide more details on how generation benchmarks (e.g., `gsm8k`) are evaluated? Any insights would be greatly appreciated!
> 
> Thanks in advance!

Dear @f14-bertolotti ,

Your replication code is similar to ours. Additionally, we attempted batch generation. Could you please let us know which GPU you used and how much time it took in total to reproduce your results? We would greatly appreciate it if you could share this information with us.

Best regards,
Jorji

**Comment by NieShenRuc** [comment]
> Dear [@Monohydroxides](https://github.com/Monohydroxides),
> 
> I would like to know if your internal evaluation library supports batch processing for conditional generation evaluation. Additionally, could you specify which GPU is used for this process? What is the maximum duration for an evaluation on a benchmark? Thank you for taking the time to answer my questions.
> 
> Best regards, Jorji

Thank you for your attention!

We are not using batch processing, which means the inference batch_size=1. This is to ensure that the answer length for each question is the same (e.g., 1024 for the Base model on GSM8K). I have not tested the scenario with a batch size >1, which might require careful handling of padding. This remains an open question. If the batch size > 1, when the prompt lengths differ, it might be reasonable to right-align the prompts (i.e., padding the mask token on the left side of shorter prompts). I must emphasize again that I have not tested batch processing. We highly encourage everyone to improve all aspects of LLaDA and share their experimental records and insights. Rigorous contributions will be shared with others.

Considering time constraints, we conducted tests using 8 A100 GPUs. For GSM8K, the Instruct model took approximately 2 hours and 30 minutes (answer_length=256, steps=256), while the Base model took about 13 hours (answer_length=1024, steps=1024).

**Comment by NieShenRuc** [comment]
> **Hi [@Monohydroxides](https://github.com/Monohydroxides),**
> 
> First off, great work!
> 
> I'm trying to reproduce the results from your paper using `lm-evaluation-harness`, specifically focusing on **Table 6 (page 19)** with **"Randomly remasking"** and **LLaDA 8B Base** (which has a reported value of **52.3**).
> 
> However, when I run my implementation, I only obtain a score of around **20** using `lm-evaluation-harness`.
> 
> My code is based on your [`generate.py`](https://github.com/ML-GSAI/LLaDA/blob/main/generate.py). Below is a minimal version that should be reproducible as long as `generate.py` is available:
> 
> from generate import generate
> import transformers, tqdm, pprint, lm_eval, torch
> 
> class RandomMaskingLLaDA(lm_eval.api.model.LM):
>     def __init__(self):
>         super().__init__()
> 
>         self.model = transformers.AutoModelForCausalLM.from_pretrained(
>             "GSAI-ML/LLaDA-8B-Base", 
>             trust_remote_code=True, 
>             torch_dtype=torch.bfloat16, 
>         ).to("cuda:0")
>         self.tokenizer = transformers.AutoTokenizer.from_pretrained(
>             "GSAI-ML/LLaDA-8B-Base", 
>             trust_remote_code=True
>         )
> 
>         self.device = "cuda:0"
>         self.mask_id = 126336
>         self.model.eval()
> 
>     def loglikelihood(*args, **kwargs): raise NotImplemented() 
>     def loglikelihood_rolling(*args, **kwargs): raise NotImplemented() 
> 
>     @torch.no_grad()
>     def generate_until(self, requests):
>         results = []
>         for i,request in enumerate(bar:=tqdm.tqdm(requests)):
>             bar.set_description(f"{i}")
>             source_ten = torch.tensor(self.tokenizer(request.args[0])["input_ids"]).unsqueeze(0)
> 
>             result = generate(
>                 self.model,
>                 prompt=source_ten,
>                 steps=256,
>                 gen_length=512,
>                 block_length=512,
>                 temperature=0.0,
>                 cfg_scale=0.0,
>                 remasking="random",
>                 mask_id=126336,
>             )
>             results.append(self.tokenizer.decode(result[0,-512:]))
> 
>         return results
> 
> if __name__ == "__main__":
>     results = lm_eval.simple_evaluate(
>         model = RandomMaskingLLaDA(),
>         tasks=["gsm8k"],
>         limit = None, # no. requests
>         num_fewshot = 4,
>     )
>     with open("results.txt", "w") as f:
>         pprint.pprint(results, stream=f)
> I completely understand and respect your decision not to share code related to generation. However, could you provide more details on how generation benchmarks (e.g., `gsm8k`) are evaluated? Any insights would be greatly appreciated!
> 
> Thanks in advance!

Thank you for your attention! This is indeed a very peculiar result. I understand that different testing frameworks may yield varying outcomes due to differences in prompt settings, special token configurations, and other factors. However, such a significant discrepancy seems unreasonable. I will attempt to test GSM8K using lm-eval, but I currently have many urgent tasks at hand, so this matter will not be a high priority for me at the moment.

**Comment by f14-bertolotti** [comment]
@CoderChen01 Why did you close the issue? 

Were you able to reproduce their result? If so, would you mind sharing the the `generate` function?

**Comment by maomaocun** [comment]
I can reproduce it according to the settings in the appendix of the paper. I have reproduced the results of the instruct model on the gsm8k mmlupro dataset, and the error is within 0.5%

**Comment by NieShenRuc** [comment]
@f14-bertolotti So sorry for the late reply. After receiving your feedback, I checked my experiment logs and code, and discovered some bugs in the initial experiments related to using 'Randomly remasking'. There were differences between my initial implementation and the code I made public on GitHub (for the sake of brevity, the latter removed many redundant code sections)."

To be specific, in my initial implementation when calculating remasking confidence, I added Gumbel noise to the logits. Theoretically, temperature=0 equals Lowest confidence remasking while temperature=infinity equals Randomly remasking - this is common practice in image masked diffusion models like [TiTOK](https://github.com/bytedance/1d-tokenizer/blob/main/modeling/maskgit.py#L182). So when testing Randomly remasking, I set the temperature to a very large value. I've noticed this approach causes numerical overflow in the calculated confidence values, ultimately producing results similar to left-to-right autoregressive generation.

After fixing this bug, I got the final results similar to yours. Fortunately, all the results in our main text were obtained using the "Lowest confidence remasking" method, so this bug only affects Table 6.

I really appreciate your discovery of this bug. I'll upload a new version of the paper as soon as possible to correct this error. I'll also express my gratitude for your contribution on the official website of LLaDA. If you have any other questions, I sincerely welcome further discussions.

BTW, if you've successfully reproduced the results of LLaDA-Base and LLaDA-Instruct using the lm-eval library, would you mind submitting a pull request?

**Comment by NieShenRuc** [comment]
> I can reproduce it according to the settings in the appendix of the paper. I have reproduced the results of the instruct model on the gsm8k mmlupro dataset, and the error is within 0.5%

@maomaocun Thank you for your interest in our work! I'm glad to see that you have successfully reproduced LLaDA. I sincerely wonder if you could share your test code with me. You can do this by submitting a pull request or sending an email to me at nieshen@ruc.edu.cn.

I've been trying to reproduce LLaDA using the lm-eval library recently. I've successfully reproduced the results of LLaDA-Base, but for the Instruct model, I'm not sure how to add the LLaDA-Instruct chat template to the examples during the few-shot testing.

Since I have many other things to deal with, I would be extremely grateful if you could help me out. I will express my gratitude to you on the official website of LLaDA or cite your project.

**Comment by f14-bertolotti** [comment]
@NieShenRuc 

Thank you for looking into the eval.

> BTW, if you've successfully reproduced the results of LLaDA-Base and LLaDA-Instruct using the lm-eval library, would you mind submitting a pull request?

I only worked on LLaDA-Base with random remasking (and it's been a while since I wrote the code). I can look into it — if I'm successful, I'll submit a PR.
